### PR TITLE
Text background color

### LINKED ISSUE #130: Support text background color
I should be possible to define the background color of any selection within the editor.
Selection rectangle should be drawn over the background rectangle

### PR DESCRIPTION
**Pull request by amantoux**
Full fledged background color supported

- [x] Background rectangle is painted behind the selection rectangle
- [x] Provide a button in the toolbar to display the current background color and select a new one
- [x] Update example
- [x] Update HTML codec
- [ ] Update demo app on website

Fixes #130

### CODE REVIEW

**Comment by Amir-P** [comment]
Although I understand that it's still a draft, but I noticed the color picker in the checklist and wanted to share my opinion on it. I am opposed to relying on third-party packages for color picking. I find it better to let the user provide the button and the desired behavior for this such features. @amantoux

**Comment by amantoux** [comment]
> Although I understand that it's still a draft, but I noticed the color picker in the checklist and wanted to share my opinion on it. I am opposed to relying on third-party packages for color picking. I find it better to let the user provide the button and the desired behavior for this such features. @amantoux

no intention to rely on third party here

**Comment by amantoux** [comment]
> Although I understand that it's still a draft, but I noticed the color picker in the checklist and wanted to share my opinion on it. I am opposed to relying on third-party packages for color picking. I find it better to let the user provide the button and the desired behavior for this such features. @amantoux

We should put the editor toolbar in the example section I think

**Comment by Amir-P** [comment]
> (I've moved it to the example part in this PR following your statement on the fact that user's should implement their own color picker)

I am also in favor of having a simple toolbar as part of the editor for convinience and to cover the basic usage.

What I was trying to say was to not depend on a specific color picking library. First thing that comes to my mind is having a callback for the background color button that is provided by the end user. @amantoux

**Comment by amantoux** [comment]
> 

> > (I've moved it to the example part in this PR following your statement on the fact that user's should implement their own color picker)
> 
> I am also in favor of having a simple toolbar as part of the editor for convinience and to cover the basic usage.
> 
> What I was trying to say was to not depend on a specific color picking library. First thing that comes to my mind is having a callback for the background color button that is provided by the end user. @amantoux

Great!

I also had the callback in mind with an implementation in the example

Thanks @Amir-P

**Comment by amantoux** [comment]
@Amir-P I think I'm done

**Comment by amantoux** [comment]
> > I am also in favor of having a simple toolbar as part of the editor for convinience and to cover the basic usage.
> > What I was trying to say was to not depend on a specific color picking library. First thing that comes to my mind is having a callback for the background color button that is provided by the end user. @amantoux
> 
> Great!
> 
> I also had the callback in mind with an implementation in the example
> 
> Thanks @Amir-P

Finally, I opted for a default implementation that can be overridden by user

**Review by Amir-P — COMMENTED** [review]
Thanks. LGTM. Just a small comment.

File: packages/parchment/lib/src/document/attributes.dart [hunk 1]
```diff
@@ -438,12 +438,20 @@ class _InlineCodeAttribute extends ParchmentAttribute<bool> {
 /// [ParchmentAttribute.backgroundColor] instead.
 class BackgroundColorAttributeBuilder extends ParchmentAttributeBuilder<int> {
   static const _kBgColor = 'bg';
+  static const _transparentColor = 0;
 
   const BackgroundColorAttributeBuilder._()
       : super._(_kBgColor, ParchmentAttributeScope.inline);
 
   ParchmentAttribute<int> fromString(String value) =>
```
[the remark this answers is not in the document]
Comment (amantoux) [reply packages/parchment/lib/src/document/attributes.dart:446]: `withColor` is to be consistent with `ParchmentAttribute<int> withLevel(int level)`
Functionally we are assigning a color

[the remark this answers is not in the document]
Comment (Amir-P) [reply packages/parchment/lib/src/document/attributes.dart:445]: I meant the `fromString`. Not the constant.😁 @amantoux

**Comment by amantoux** [comment]
> We have conflicts after merging #128 @amantoux

Yes, it is expected. On it

**Pull request merged**